Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx

```diff
@@ -0,0 +1,427 @@
+  //}
+  /// else {
+  //    if(mTrapConfig != nullptr){
+  //      delete mTrapConfig;
+  //    mTrapConfig = nullptr;
+  //  }
+  // mTrapConfig->LOADFROMDISKDEFAULT();
+  LOG(info) << "looking for TRAPconfig " << mTrapConfigName << __FILE__ << ":" << __LINE__;
+  LOG(warn) << "No TRAPconfig entry found for name of " << mTrapConfigName;
+  LOG(info) << "looking for TRAPconfig " << mTrapConfigName << __FILE__ << ":" << __LINE__;
+  //  }
+}
+
+void TRDDPLTrapSimulatorTask::init(o2::framework::InitContext& ic)
+{
+  //               LOG(debug) << "Input data file is : " << ic.options().get<std::string>("simdatasrc");
+  //               LOG(info) << "simSm is : " << ic.options().get<int>("simSM");
+  LOG(debug1) << "TRD Trap Simulator Device with pid of : " << ::getpid();
+  mfeeparam = FeeParam::instance();
+  mPrintTrackletOptions = ic.options().get<int>("printtracklets");
+  mDrawTrackletOptions = ic.options().get<int>("drawtracklets");
+  mShowTrackletStats = ic.options().get<int>("show-trd-trackletstats");
+  mTrapConfigName = ic.options().get<std::string>("trapconfig");
+  LOG(info) << "disable trap simualtor is : " << mDisableTrapSimulation;
+  LOG(info) << "Trap Simulator Device initialising with trap config of : " << mTrapConfigName;
+  //  if(mDisableTrapSimulation){
+  //  //now get a trapconfig to work with.
+  //    LOG(warn) << "You elected to not do a trap chip simulation and hence no trapconfig was saught";
+  //  }
+  //  else{
+  getTrapConfig();
+  LOG(info) << "Trap Simulator Device initialised ... ";
+  //  }
+}
+
+void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
+{
+  LOG(info) << "TRD Trap Simulator Device running over incoming message ...";
+  if (mDisableTrapSimulation) {
+    LOG(warn) << " you elected to not do a trap chip simulation";
+    return;
+  }
+  // get the relevant inputs for the TrapSimulator
+  auto digits = pc.inputs().get<std::vector<o2::trd::Digit>>("digitinput");
+  //auto mMCLabels = pc.inputs().get<o2::dataformats::MCTruthContainer<o2::trd::MCLabel>*>("labelinput");
+  //auto mTriggerRecords = pc.inputs().get<std::vector<o2::trd::TriggerRecord>>("triggerrecords");
+
+  LOG(debug) << "Read in Digits with size of : " << digits.size();
+
+  //set up structures to hold the returning tracklets.
+  std::vector<Tracklet> mMCMTracklets; //vector to store the retrieved tracklets from an mcm
+  std::vector<Tracklet> mMCMTrackletsAccum;
+  mMCMTracklets.reserve(30);
+  mMCMTrackletsAccum.reserve(digits.size() / 3); //attempt to a. conserve mem, b. stop a vector resize
+
+  //TODO we can ignore time, and use the triggerrecords as to defined subsets to sort. TriggerRecord has the time in it for all the digits, by creation.
+  //
+  std::chrono::duration<double> sortingtime{0}; ///< full timer
+  auto sortstart = std::chrono::high_resolution_clock::now();
+
+  std::sort(digits.begin(), digits.end(), DigitSortComparator);
```

Comment:
how about doing the sorting on an index instead of moving around the digit objects?

Reply:
already done, PR should come today. Oone for the lack of moving objects, and also more importantly the conservation of the labels to digit mapping. Additionally only need to sort between trigger records and dont need to sort on time as the time part of the sorting is never hit due to the construction of the digits vector.